### PR TITLE
feat(CapMan): AllocationPolicy for Errors

### PR DESCRIPTION
**Pull request by volokluev**
This PR introduces the first iteration of the allocation policy for the errors storage

### The Policy

For every organization, keep track of how many bytes they've scanned in the last 10 minutes, when they reach a limit, throttle them to one thread

### How to turn it on/off

In runtime config:

```
# completely bypasses the code path
SET ErrorsAllocationPolicy.is_active 0
```

```
# code runs and emits metrics but no action is taken
SET ErrorsAllocationPolicy.is_enforced 0
```

```
# how many bytes the org can scan in ten minutes before being throttled
SET ErrorsAllocationPolicy.org_scan_limit 42069
```

### Intent

This is not meant to be the final policy, this is just a quick way to get something working end to end and then iterate on it. Most of this code will be rewritten

### CODE REVIEW

File: snuba/query/allocation_policies/errors_allocation_policy.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+                10000,
+            )
+
+            timestamp, granted_quotas = self._rate_limiter.check_within_quotas(
+                [
+                    RequestedQuota(
+                        self.rate_limit_prefix,
+                        # request a big number because we don't know how much we actually
+                        # will use in this query. this doesn't use up any quota, we just want to know how much is left
+                        int(1e10),
```
Comment (dbanda) [inline snuba/query/allocation_policies/errors_allocation_policy.py:110]: What are the units for this value? Is it bytes scanned?

File: snuba/query/allocation_policies/errors_allocation_policy.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                            granularity_seconds=self.WINDOW_GRANULARITY_SECONDS,
+                            limit=int(org_scan_limit),  # type: ignore
+                            prefix_override=f"{self.rate_limit_prefix}-organization_id-{tenant_ids['organization_id']}",
+                        )
+                    ],
+                )
+            ],
+            grants=[
+                GrantedQuota(
+                    "{self.rate_limit_prefix}-organization_id-{tenant_ids['organization_id']}",
```
Comment (rahul-kumar-saini) [inline snuba/query/allocation_policies/errors_allocation_policy.py:180]:
```suggestion
                    f"{self.rate_limit_prefix}-organization_id-{tenant_ids['organization_id']}",
```

**Review by rahul-kumar-saini — COMMENTED** [review]
Just some nits, also how easy/clean would it be to pass quota data down to querylog in one of the existing json blobs like clickhouse stats?

File: snuba/query/allocation_policies/__init__.py [hunk 3]
```diff
@@ -162,7 +167,15 @@ def from_kwargs(cls, **kwargs: str) -> "AllocationPolicy":
         )
 
     def get_quota_allowance(self, tenant_ids: dict[str, str | int]) -> QuotaAllowance:
-        allowance = self._get_quota_allowance(tenant_ids)
+        try:
+            allowance = self._get_quota_allowance(tenant_ids)
+        except Exception:
+            logger.exception(
+                "Allocation policy failed to get quota allowance, this is a bug, fix it"
+            )
```
[the remark this answers is not in the document]
Comment (volokluev) [reply snuba/query/allocation_policies/__init__.py:175]: logger.exception should propagate to sentry and logs

[the remark this answers is not in the document]
Comment (rahul-kumar-saini) [reply snuba/query/allocation_policies/__init__.py:175]: I meant the message from the original exception that was caught, this looks like it'd just say "Allocation policy failed" without any context no?

[the remark this answers is not in the document]
Comment (volokluev) [reply snuba/query/allocation_policies/__init__.py:175]: I understand your question now. `logging.exception` captures the exception information as well, example:

```
>>> import logging
>>> try:
...     raise Exception("This is an error")
... except Exception:
...     logging.exception("We messed up")
...
ERROR:root:We messed up
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
Exception: This is an error
```

File: snuba/query/allocation_policies/errors_allocation_policy.py (hunk 1 above)
Comment (volokluev) [reply snuba/query/allocation_policies/errors_allocation_policy.py:116]: good question. Yes but I will clarify that

**Comment by volokluev** [comment]
@rahul-kumar-saini Addressed all comments, also policy decisions are now captured in `stats` in the querylog

File: snuba/query/allocation_policies/__init__.py [hunk 4]
```diff
@@ -162,7 +173,15 @@ def from_kwargs(cls, **kwargs: str) -> "AllocationPolicy":
         )
 
     def get_quota_allowance(self, tenant_ids: dict[str, str | int]) -> QuotaAllowance:
-        allowance = self._get_quota_allowance(tenant_ids)
+        try:
+            allowance = self._get_quota_allowance(tenant_ids)
+        except Exception:
+            logger.exception(
+                "Allocation policy failed to get quota allowance, this is a bug, fix it"
```
Comment (evanh) [inline snuba/query/allocation_policies/__init__.py:180]: It's probably worth capturing the tenant_ids here too.

File: snuba/query/allocation_policies/errors_allocation_policy.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    RequestedQuota,
+)
+from snuba.utils.metrics.wrapper import MetricsWrapper
+
+metrics = MetricsWrapper(environment.metrics, "errors_allocation_policy")
+
+
+# A hardcoded list of referrers which do not have an organization_id associated with them
+# purposefully not in config because we don't want that to be easily changeable
+_ORG_LESS_REFERRERS = set(
```
Comment (evanh) [inline snuba/query/allocation_policies/errors_allocation_policy.py:29]: Why have this list? Why not skip any query that doesn't have an `organization_id` in it?

**Review by rahul-kumar-saini — APPROVED** [review]

File: snuba/query/allocation_policies/errors_allocation_policy.py (hunk 5 above)
Comment (volokluev) [reply snuba/query/allocation_policies/errors_allocation_policy.py:29]: some of them are valid, e.g. subscriptions. Some of them are maybe valid but I don't have enough knowledge. They don't constitute even a percent of traffic but I think you're right that we should make this list go away eventually

**Pull request merged**